cumulus_library/studies/core/soe_sequence_of_events.sql

```diff
@@ -0,0 +1,343 @@
+-- SOE "Sequence of Events" links FHIR resources based on encounter reference
+-- and FHIR resource datetime "period". Ideally, every FHIR resource references
+-- an encounter. The primary function of the SOE tables and counts is QA/Verification.
+--
+-- ###########################################################################
+-- Encounter
+-- Encounter specifies the start/end period from which other FHIR resources can
+-- be mapped. This is useful when FHIR Resource.encounter.reference is missing.
+--
+-- # length of stay calculations
+-- From Karen Olson
+--  LOS= discharge date/time - admit date/time
+--  EDLOS= checkout date/time - check-in date/time
+--  check-in date/time generally = admit date/time, sometimes it's off a little
+--  Care_class= Emergency is ED only
+--  An encounter with care_class= Inpatient or Observation can include time in the ED
+--  If patient was admitted from the ED, THEN only encounter.
+--
+create table core__soe as
+with rawdata as
+(
+    select distinct
+        date(from_iso8601_timestamp(e.period."start"))  AS enc_start_date,
+        date(from_iso8601_timestamp(e.period."end"))    AS enc_end_date,
+        cast(from_iso8601_timestamp(e.period."start") as timestamp) AS enc_start_datetime,
+        cast(from_iso8601_timestamp(e.period."end") as timestamp)   AS enc_end_datetime,
+        e.class as enc_class,
+        subject.reference as subject_ref,
+        concat('Encounter/', e.id) AS encounter_ref
+    FROM encounter AS e
+)
+select distinct
+    rawdata.*,
+    date_diff('hour', enc_start_datetime, enc_end_datetime) as los_hours,
+    date_diff('day', enc_start_date, enc_end_date) as los_days
+from rawdata
+where enc_start_date between date('2016-01-01') and current_date
+;
+
+-- ###########################################################################
+-- Condition
+
+create table core__soe_cond as
+with rawdata as
+(
+    SELECT distinct
+        cast(from_iso8601_timestamp(c.recordedDate) as timestamp)  AS cond_recorded_datetime,
+        cast(from_iso8601_timestamp(c.onsetDateTime) as timestamp) AS cond_onset_datetime,
+        c.subject.reference     AS subject_ref,
+        c.encounter.reference   AS encounter_ref,
+        c.category,
+        concat('Condition/', c.id) AS condition_ref
+    FROM condition AS c
+),
+cond as
+(
+    select distinct rawdata.*, category_row
+    from rawdata,
+        unnest(category) AS t1 (category_coding),
+        unnest(category_coding.coding) AS t2 (category_row)
```

Comment:
This unnest means we need to go through the various investigations of 'how does this come back in epic vs cerner vs others' - we may have to move this statement into a table builder.